consensus/pacemaker/module.go

```diff
@@ -243,6 +265,87 @@ func (m *pacemaker) NewHeight() {
 		)
 }
 
+// This is called each time there is a NewRound message received by the leader from replicas
+// With the introduction of MinBlockTimeMsec delay, multiple concurrent calls may happen
+// It makes sure that:
+//   - Block proposal is made by only one of the possible `HotstuffLeaderMessageHandler.HandleNewRoundMessage()` concurrent (because delayed) calls
+//   - If the timer expires, the first call to this method will trigger the block proposal
+//   - If a late message is received after a block is proposed by another call, the late message is discarded
+//   - Reads and affectations to pacemaker.latestPrepareRequest state are protected by a mutex
+func (m *pacemaker) ProcessDelayedBlockPrepare(currentHeight uint64) chan bool {
+	m.latestPrepareRequest.m.Lock()
+	defer m.latestPrepareRequest.m.Unlock()
+
+	// Prepare channel for the the current request
+	ch := make(chan bool)
+
+	// First time to build a block for current height, cancel previous timer if any
+	if m.latestPrepareRequest.delayedHeight < currentHeight {
+		// Discard the request building the old height
+		if m.latestPrepareRequest.ch != nil {
+			m.latestPrepareRequest.ch <- false
+		}
+		// Discard the timer for the old height
+		if m.latestPrepareRequest.cancelFunc != nil {
+			m.latestPrepareRequest.cancelFunc()
+		}
+
+		m.latestPrepareRequest.ch = ch
+		m.latestPrepareRequest.blockProposed = false
```

Comment:
We should emphaize that all of these are mutually exclusive scenarios/triggers for block proposal. Is that correct?